dotCMS/src/main/java/com/dotcms/api/content/VanityUrlFactoryImpl.java

```diff
@@ -0,0 +1,226 @@
+package com.dotcms.api.content;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.dotcms.cache.VanityUrlCache;
+import com.dotcms.content.model.VanityUrl;
+import com.dotcms.contenttype.model.type.BaseContentType;
+import com.dotcms.contenttype.model.type.ContentType;
+import com.dotcms.util.VanityUrlUtil;
+import com.dotmarketing.beans.Host;
+import com.dotmarketing.beans.Identifier;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.CacheLocator;
+import com.dotmarketing.business.DotStateException;
+import com.dotmarketing.exception.DotDataException;
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.exception.DotSecurityException;
+import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
+import com.dotmarketing.portlets.contentlet.business.DotContentletStateException;
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+import com.dotmarketing.portlets.folders.model.Folder;
+import com.dotmarketing.util.Config;
+import com.dotmarketing.util.InodeUtils;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.UtilMethods;
+import com.liferay.portal.model.User;
+
+/**
+ * Implementation class for the {@link VanityUrlFactory}.
+ * @author oswaldogallango
+ *
+ */
+public class VanityUrlFactoryImpl implements VanityUrlFactory {
+	final ContentletAPI contentletAPI = APILocator.getContentletAPI();
+	final VanityUrlCache vanityURLCache = CacheLocator.getVanityURLCache();
+
+	@Override
+	public List<VanityUrl> getAllVanityUrls(User user) throws DotDataException, DotSecurityException {
+		List<VanityUrl> results = new ArrayList<VanityUrl>();
+		List<Contentlet> contentResults = contentletAPI.search("+baseType:"+BaseContentType.VANITY_URL.getType()+" +working:true", 0, 0, "", user, false);
+		contentResults.stream().forEach((Contentlet con) ->{
+			results.add(fromContentlet(con));
+		});
+
+		return results;
+	}
+
+	@Override
+	public List<VanityUrl> getActiveVanityUrls(User user) throws DotDataException, DotSecurityException {
+		List<VanityUrl> results = new ArrayList<VanityUrl>();
+		List<Contentlet> contentResults = contentletAPI.search("+baseType:"+BaseContentType.VANITY_URL.getType()+" +live:true +deleted:false", 0, 0, "", user, false);
+		contentResults.stream().forEach((Contentlet con) -> {
+			VanityUrl vanityUrl = fromContentlet(con);
+			addToVanityURLCache(vanityUrl);
+			results.add(vanityUrl);
+		});
+		
+		return results;
+	}
+
+	@Override
+	public VanityUrl getVanityUrlByURI(String uri, Host host, long languageId, User user, boolean live) 
```

Comment:
So, I think this is the method that needs to do the cache lookup.  

If there is a match in cache, return it.

If there not a match in cache, we will do the content lookup (I think I'd prefer a db lookup and then a contentAPI.find() vs relying on the index).

If there is no match found in the DB, we will need to add a 404CacheEntry so following lookups will not incur a db hit.  

Either way, put the 404CacheEntry or the vanity url in cache and return it.